cli/cmd/devtool/start.go

```diff
@@ -0,0 +1,759 @@
+		return fmt.Errorf("docker not available: found errors: %v", se)
+	}
+
+	return nil
+}
+
+func checkRillRepo(ctx context.Context) error {
+	_, err := os.Stat(".git")
+	if err != nil {
+		return fmt.Errorf("you must run `rill devtool` from the root of the rill repository")
+	}
+
+	out, err := newCmd(ctx, "git", "remote", "get-url", "origin").Output()
+	if err != nil {
+		return fmt.Errorf("error executing the 'git remote get-url origin' command: %w", err)
+	}
+
+	repo := strings.TrimSpace(string(out))
+	rillRepo := "https://github.com/rilldata/rill.git"
+	if repo != rillRepo {
+		return fmt.Errorf("you must run `rill devtool` from the rill repository (expected remote %q, got %q)", rillRepo, repo)
+	}
+
+	return nil
+}
+
+type servicesCfg struct {
+	admin   bool
+	deps    bool
+	runtime bool
+	ui      bool
+	only    []string
+	except  []string
+}
+
+func (s *servicesCfg) addFlags(cmd *cobra.Command) {
+	cmd.Flags().StringSliceVar(&s.only, "only", []string{}, "Only start the listed services (options: admin, deps, runtime, ui)")
+	cmd.Flags().StringSliceVar(&s.except, "except", []string{}, "Start all except the listed services (options: admin, deps, runtime, ui)")
+}
+
+func (s *servicesCfg) parse() error {
+	if len(s.only) > 0 && len(s.except) > 0 {
+		return errors.New("cannot use both --only and --except")
+	}
+
+	vals := s.except
+	def := true
+	if len(s.only) > 0 {
+		vals = s.only
+		def = false
+	}
+
+	s.admin = def
+	s.deps = def
+	s.runtime = def
+	s.ui = def
+
+	for _, v := range vals {
+		switch v {
+		case "admin":
+			s.admin = !def
+		case "deps":
+			s.deps = !def
+		case "runtime":
+			s.runtime = !def
+		case "ui":
+			s.ui = !def
+		default:
+			return fmt.Errorf("invalid service %q", v)
+		}
+	}
+
+	return nil
+}
+
+type cloud struct{}
+
+func (s cloud) start(ctx context.Context, ch *cmdutil.Helper, verbose, reset, refreshDotenv bool, services *servicesCfg) error {
+	if reset {
+		err := s.resetState(ctx)
+		if err != nil {
+			return fmt.Errorf("reset cloud deps: %w", err)
+		}
+		logInfo.Printf("Reset cloud dependencies\n")
+	}
+
+	if refreshDotenv {
+		err := downloadDotenv(ctx, "cloud")
+		if err != nil {
+			return fmt.Errorf("failed to refresh .env: %w", err)
+		}
+		logInfo.Printf("Refreshed .env\n")
+	}
+
+	// Validate the .env file is well-formed.
+	err := checkDotenv()
+	if err != nil {
+		return err
+	}
+	_, err = godotenv.Read()
+	if err != nil {
+		return fmt.Errorf("error parsing .env: %w", err)
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	if services.deps {
+		g.Go(func() error { return s.runDeps(ctx, verbose) })
+	}
+
+	depsReadyCh := make(chan struct{})
+	g.Go(func() error {
+		if services.deps {
+			err := s.awaitPostgres(ctx)
+			if err != nil {
+				return err
+			}
+			err = s.awaitRedis(ctx)
+			if err != nil {
+				return err
+			}
+		}
+		close(depsReadyCh)
+		return nil
+	})
+
+	if services.admin {
+		g.Go(func() error {
+			select {
+			case <-depsReadyCh:
```

Comment:
Didn't do this because a) the `npm install` is often slow and therefore nice to start as early as possible, b) seems cleaner to make sure `g.Wait()` is called quickly (to correctly and quickly propagate errors), which also keeps each goroutine independently responsible for ensuring its preconditions.

I just added a util function `awaitClose` to clean up the various `select` statements.